### ISSUE TITLE
Server restart needed to see changes in development? (v 1.0.0 / v 5.0.0)

### ISSUE THREAD

**Issue by Jamevid** [sferik/rails_admin#2726]
Before v1.0.0 changes in "rails_admin do" - block  would show up in development directly... by browser refresh. 

Now in v 1.0.0 and rails 5.0.0 a server restart is required for rails_admin changes to take effect. 

Should it be like that?

Thanks for any input! :-)

**Comment by niv** [comment]
I'm seeing the same issue!

Additionally, when using the pundit extension, any model changes trigger a controller reload, which evicts the Pundit include (and thus all following requests fail because policy() isnt in scope). This might be related, so I'm not opening a new issue with this immediately.

**Comment by patleb** [comment]
correction [here](https://github.com/patleb/rails_admin/blob/master/lib/rails_admin/engine.rb#L38)

**Comment by sfcgeorge** [comment]
I noticed the same issue, as the rails_admin DSL is pretty tricky it gets quite frustrating reloading the server after every model setting change. 

The commit @patleb is referring to is https://github.com/patleb/rails_admin/commit/f4abececc1f1a2b9659244313aa212c72b0d32ba

For those wanting to monkeypatch this themselves until fixed, add the following to the top of your `config/initializers/rails_admin.rb`

```ruby 
RailsAdmin::Engine.config.to_prepare do
  return unless Rails.env.development?
  RailsAdmin::Config.reset
  load "#{Rails.root}/config/initializers/rails_admin.rb"
end

# ...
```

@sferik would a pull request of the above commit be helpful? I don't really understand what `config.to_prepare` is doing or when it runs so there may be a more efficient place to put the `RailsAdmin::Config.reset`.

**Comment by vergenzt** [comment]
FWIW I've found that config does partially reload for me, but it only does a sort of hybrid *union*-reload. My `rails_admin` model blocks get *appended* to previously defined blocks rather than replacing the model-level ones (which can result in the appearance of not reloading at all), which makes sense given the way `lazy_model` is implemented.

Somehow `LazyModel#target` needs to purge stale blocks from previous model implementations when evaluating.

**Comment by sfcgeorge** [comment]
@vergenzt yep that's what I've noticed. I only have config in the models themselves and even then, each change seems to append rather than replace, leading to weird errors where the highlighted source is different to the error message because it's stale. You could try the fix above and see if that helps your setup too.

**Comment by dkobia** [comment]
@sfcgeorge a slight change to your code to prevent a `LocalJumpError: unexpected return` error in production:

```
RailsAdmin::Engine.config.to_prepare do
  if Rails.env.development?
    RailsAdmin::Config.reset
    load "#{Rails.root}/config/initializers/rails_admin.rb"
  end
end
```

**Comment by KenneyE** [comment]
The above solutions still don't seem to work for me. I'm using v1.4.2 with Rails 5.1. 
I have the following at the top of my `rails_admin.rb` intializer:
```
RailsAdmin::Engine.config.to_prepare do
  return unless Rails.env.development?
  RailsAdmin::Config.reset
  load "#{Rails.root}/config/initializers/rails_admin.rb"
end
```
Like other people have seen, I'm getting a partial reload where the config is being extended rather than replaced. Is anyone able to confirm that the above solution has worked for them?

**Comment by sfcgeorge** [comment]
Nope, a restart still needs to be required in many cases. Frustrating for big slow apps. Since none of the patches work reliably it'll be hard to debug whether your change is wrong, or it just hasn't reloaded correctly. So I'd recommend always restarting until theres a real fix.

**Comment by yutoji** [comment]
In order to reset only after the file change is detected: 
Add this file under the `config/initializers/` dir.

```
# See also: https://github.com/sferik/rails_admin/wiki/How-to:-Reloading-RailsAdmin-Config-Automatically#alternative-method-for-the-railsadmin-versions-that-does-not-extend-applicationcontroller
Rails.application.config.to_prepare do
  RailsAdmin::ApplicationController.class_eval do
    before_action :reload_rails_admin, if: :reload_rails_admin? # Reloading RailsAdmin Config Automatically

    RAILS_ADMIN_FILE_PATH = "#{Rails.root}/config/initializers/rails_admin.rb"

    class << self
      def fetch_rails_admin_file_updated_time
        File::Stat.new(RAILS_ADMIN_FILE_PATH).mtime
      end
      attr_accessor :__file_timestamp
    end
    @__file_timestamp = fetch_rails_admin_file_updated_time

    def reload_rails_admin
      RailsAdmin::Config.reset

      load(RAILS_ADMIN_FILE_PATH)
    end

    def reload_rails_admin?
      return false unless Rails.env.development?
      begin
        file_updated_at = RailsAdmin::ApplicationController.fetch_rails_admin_file_updated_time
        has_changed = file_updated_at != self.class.__file_timestamp
        return has_changed
      ensure
        self.class.__file_timestamp = file_updated_at
      end
    end
  end
end
```

**Comment by dfabreguette-ap** [comment]
As I defined a "rails_admin do (...) end" specific section in my models,
I added a "load" over all models ruby files in app/models/...rb"

```ruby
Rails.application.config.to_prepare do
  RailsAdmin::ApplicationController.class_eval do
    before_action :reload_rails_admin, if: :reload_rails_admin? # Reloading RailsAdmin Config Automatically

    RAILS_ADMIN_FILE_PATH = "#{Rails.root}/config/initializers/rails_admin.rb"

    class << self
      def fetch_rails_admin_file_updated_time
        File::Stat.new(RAILS_ADMIN_FILE_PATH).mtime
      end
      attr_accessor :__file_timestamp
    end
    @__file_timestamp = fetch_rails_admin_file_updated_time

    def reload_rails_admin
      RailsAdmin::Config.reset

      load(RAILS_ADMIN_FILE_PATH)
      # Loop over app/models ruby files and load them individually
      Dir["#{Rails.root}/app/models/*.rb"].each {|file|
        load(file)
      }
    end

    def reload_rails_admin?
      return false unless Rails.env.development?
      begin
        file_updated_at = RailsAdmin::ApplicationController.fetch_rails_admin_file_updated_time
        has_changed = file_updated_at != self.class.__file_timestamp
        return has_changed
      ensure
        self.class.__file_timestamp = file_updated_at
      end
    end
  end
end
```

**Comment by Syntaf** [comment]
If you're using model concerns for rails_admin configuration, I wrote a quick tutorial on how I managed to implemented dynamic reloading:

https://medium.com/@GrantMercer/dynamically-reloading-rails-admin-model-concerns-in-development-f5f6959f574b

Gist: https://gist.github.com/Syntaf/623f9a09c2ef202c129245de7f8c0855

**Comment by JoshCheek** [comment]
It took me several hours to figure this out, but `RailsAdmin::Config.reset` isn't enough. Several other classes store state that needs to get reset, too. See code sample for what I'm currently doing.  **Note** that all our config happens in config/initializers/rails_admin.rb, we don't define any in the controllers themselves, if you do, then you might need to incorporate ideas from some of the above comments. I'll update this comment if I need to modify this solution:

```ruby
class ApplicationController < ActionController::Base
  # In development, add a before_action to reload rails admin so we don't have to restart Spring between changes
  # Modified the solution proposed here: https://github.com/sferik/rails_admin/wiki/How-to:-Reloading-RailsAdmin-Config-Automatically
  # There is discussion about it here:   https://github.com/sferik/rails_admin/issues/2726
  Rails.env.development? && before_action do
    next unless controller_path =~ /rails_admin/
    RailsAdmin::Config.reset
    RailsAdmin::AbstractModel.reset
    RailsAdmin::AbstractModel.reset_polymorphic_parents
    RailsAdmin::Config::Actions.reset
    load Rails.root/"config/initializers/rails_admin.rb"
  end
end
```
